### PR TITLE
Infer character encoding for ifAlias and sysLocation

### PR DESCRIPTION
**Pull request by murrant**
Tries to convert character encoding for non-UTF-8 encoded strings.
This will only work for snmp strings that are type-hinted, not ones forced to ASCII with -Oa or similar
Only works for your default charset or Windows-1251 or LATIN1.
You can set your character encoding in .env with CHARSET

**Please save us all and just use UTF-8**

Thanks to symfony/var-dumper for the hints.

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

File: LibreNMS/Util/StringHelpers.php [hunk 1]
```diff
@@ -93,4 +93,36 @@ public static function camelToTitle($string)
...
+            return (string) $converted;
+        }
+
+        if ($charset !== 'Windows-1252' && $converted = @iconv('Windows-1252', 'UTF-8', $string) !== false) {
+            return (string) $converted;
+        }
+
+        if ($charset !== 'CP850' && $converted =  @iconv('Windows-1252', 'UTF-8', $string) !== false) {
+            return (string) $converted;
+        }
```
Comment (Jellyfrog) [inline LibreNMS/Util/StringHelpers.php:122]: Test fails due to $converted is set to the result of false test.

```suggestion
        if (($converted = @iconv($charset, 'UTF-8', $string)) !== false) {
            return (string) $converted;
        }

        if ($charset !== 'Windows-1252' && ($converted = @iconv('Windows-1252', 'UTF-8', $string)) !== false) {
            return (string) $converted;
        }

        if ($charset !== 'CP850' && ($converted =  @iconv('Windows-1252', 'UTF-8', $string)) !== false) {
            return (string) $converted;
        }
```

File: LibreNMS/Util/StringHelpers.php [hunk 2]
```diff
@@ -93,4 +93,36 @@ public static function camelToTitle($string)
...
+
+        if ($charset !== 'Windows-1252' && ($converted = @iconv('Windows-1252', 'UTF-8', $string)) !== false) {
+            return (string) $converted;
+        }
+
+        if ($charset !== 'CP850' && ($converted =  @iconv('Windows-1252', 'UTF-8', $string)) !== false) {
+            return (string) $converted;
+        }
```
Comment (Jellyfrog) [inline LibreNMS/Util/StringHelpers.php:122]: these 2 can be combined?

Comment (murrant) [reply LibreNMS/Util/StringHelpers.php:122]: It is a typo

**Comment by kimhaak** [comment]
Tested and it fixed problems for me.

**Comment by kimhaak** [comment]
Seeing same issue on "sysname" - would that be considered a input like ifalias and location?

**Comment by murrant** [comment]
@kimhaak yes, feel free to send a Pull Request for that :)

**Comment by kimhaak** [comment]
> @kimhaak yes, feel free to send a Pull Request for that :)

ok, but i'll need a hint. Pushed the sysname string trough the helper function in includes/polling/core.inc.php - that helped when the poller ran. But when discovery runs, it changes from the correct string, to the wrong string, and then, back again.

Seems like discovery is asking for sysname from somewhere else (at one point at least) than includes/polling/core.inc.php



image

**Comment by murrant** [comment]
https://github.com/librenms/librenms/blob/master/includes/discovery/core.inc.php#L12

**Pull request merged**